File: src/main/scala/windowsapi/scala/User32.scala
package windowsapi.scala

import com.sun.jna.Native
import windowsapi.java

object User32 {
  case class DisplayDevice(name: String, string: String, active: Boolean, id: String, key: String)

  def enumDisplayDevices(): Iterator[DisplayDevice] =
    enumDisplayDevices(null)

  def enumDisplayDevices(displayAdapterName: String): Iterator[DisplayDevice] =
    Iterator.from(0)
      .map { n =>
        val displayDevice = new java.User32.DISPLAY_DEVICE
        val result = java.User32.INSTANCE.EnumDisplayDevices(displayAdapterName, n, displayDevice, 0)
        (result, displayDevice)
      }
      .takeWhile { case (result, _) => result }
      .map { case (_, displayDevice: java.User32.DISPLAY_DEVICE) =>
        DisplayDevice(
          name = Native.toString(displayDevice.DeviceName),
          string = Native.toString(displayDevice.DeviceString),
          active = (displayDevice.StateFlags & java.User32.DISPLAY_DEVICE_ACTIVE) != 0,
          id = Native.toString(displayDevice.DeviceID),
          key = Native.toString(displayDevice.DeviceKey),
        )
      }
}

File: windows/src/main/scala/windowsapi/java/User32.java
package windowsapi.java;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Structure;
import com.sun.jna.win32.W32APIOptions;

public interface User32 extends Library {
    User32 INSTANCE = Native.load("user32", User32.class, W32APIOptions.DEFAULT_OPTIONS);

    // https://docs.microsoft.com/en-us/windows/desktop/api/wingdi/ns-wingdi-display_devicew
    @Structure.FieldOrder({"cb", "DeviceName", "DeviceString", "StateFlags", "DeviceID", "DeviceKey"})
    class DISPLAY_DEVICE extends Structure {
        public int cb;
        public char DeviceName[] = new char[32];
        public char DeviceString[] = new char[128];
        public int StateFlags;
        public char DeviceID[] = new char[128];
        public char DeviceKey[] = new char[128];

        public DISPLAY_DEVICE() {
            cb = size();
        }
    }

    // https://docs.microsoft.com/en-us/windows/desktop/api/winuser/nf-winuser-enumdisplaydevicesw
    boolean EnumDisplayDevices(
        String lpDevice,
        int iDevNum,
        DISPLAY_DEVICE lpDisplayDevice,
        int dwFlags
    );

    int DISPLAY_DEVICE_ACTIVE = 1;
}
